File: LeetCode/leetbook/二分/寻找峰值.cpp
    #include <bits/stdc++.h>
    using namespace std;

    class Solution {
    public:
        int findPeakElement(vector<int>& nums) {
            int l, r, m;
            nums.insert(nums.begin(), INT_MIN);
            nums.push_back(INT_MAX);
            l = 1;
            r = nums.size() - 2;
            while (l < r) {
                m = l + (r-l >> 1);
                if (nums[m] < nums[m+1]) {
                    l = m+1;
                } else {
                    r = m;
                }
            }
            // while (l < r) {
            //     m = l + (r-l >> 1) + 1;
            //     if (nums[m] <= nums[m-1]) {
            //         r = m-1;
            //     } else {
            //         l = m;
            //     }
            // }
            return l-1;
        }
    };

File: LeetCode/leetbook/二分/寻找峰值.py
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        nums = [-inf] + nums + [inf]
        n = len(nums)
        l = 1
        r = n-2
        while l < r:
            mid = l + (r - l >> 1) + 1
            if nums[mid] > nums[mid-1]:
                l = mid
            else:
                r = mid-1
        return l-1